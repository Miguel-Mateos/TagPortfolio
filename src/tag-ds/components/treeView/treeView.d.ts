import React from 'react';
interface ITree {
    /**
     * Alternative className for Tree
     */
    className?: string;
    [other: string]: any;
}
export declare const TreeWrap: React.FC<ITree>;
export declare const Tree: {
    Wrap: React.FC<ITree>;
    Element: React.FC<import("./treeViewElement").ITreeElement>;
};
export {};
