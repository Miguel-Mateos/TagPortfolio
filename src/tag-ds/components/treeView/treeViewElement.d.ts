import React from 'react';
export interface ITreeElement {
    /**
     * Name of the Tree Element
     */
    name: string;
    /**
     * For disabling the Tree Element
     */
    disabled?: boolean;
    /**
     * click event on developers needs
     */
    onClick?: (name: string) => void;
    /**
     * Alternative className on Tree.Element
     */
    className?: string;
    [others: string]: any;
}
export declare const TreeViewElement: React.FC<ITreeElement>;
