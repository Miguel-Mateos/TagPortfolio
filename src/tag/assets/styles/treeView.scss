.tag-ds {
  .tree-element,
  &.tree-element {
    margin-top: 0;
    margin-bottom: 0;
    padding: 0;
    padding-left: calc(3 * var(--unit));
    margin-bottom: 8px;
    @include font(medium);
    text-transform: none;
    border: 0;
    border-left: calc(var(--unit) / 4) solid transparent;
    background-color: transparent;
    border-radius: 0;
    height: auto;
    width: auto;

    & .list-icon {
      margin-top: 0;
      margin-bottom: 0;
      padding-top: 0;
    }

    & > li {
      color: var(--neutral800);
      text-align: left;
      width: 100%;
    }

    &:hover {
      & > li {
        color: var(--secondary600);
      }
      background-color: transparent;
      border: 0;
      border-left: calc(var(--unit) / 4) solid transparent;
    }

    &:active {
      border: 0;
      border-left: calc(var(--unit) / 4) solid var(--accent500);
      & > li {
        color: var(--secondary700);
      }
    }

    &:focus {
      outline: calc(var(--unit) / 2) solid var(--secondary300);
      background-color: var(--neutral100);
      border: 0;
      border-left: calc(var(--unit) / 4) solid var(--accent500);
    }

    &:disabled {
      background-color: transparent;
      border: 0;
      border-left: calc(var(--unit) / 4) solid transparent;
      & > li {
        color: var(--neutral500);
      }
    }
  }

  .tree-element-item {
    margin-left: var(--unit);
  }
}
