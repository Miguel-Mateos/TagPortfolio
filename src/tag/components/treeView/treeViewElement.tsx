import React, { useState } from 'react';

export interface ITreeElement {
  /**
   * Name of the Tree Element
   */
  name: string
  /**
   * For disabling the Tree Element
   */
  disabled?: boolean
  /**
   * click event on developers needs
   */
  onClick?: (name: string) => void
  /**
   * Alternative className on Tree.Element
   */
  className?: string
  [others: string]: any
}


export const TreeViewElement: React.FC<ITreeElement> = ({ children, name, onClick, disabled, className, ...rest }) => {
  const [showChildren, setShowChildren] = useState<boolean>(false);

  const handleClickElement = () => {
    setShowChildren(!showChildren);
    if (onClick) onClick(name);
  };

  return (
    <>
      <button disabled={disabled} className={`list-icon tree-element ${className || ''}`}
        onClick={handleClickElement} {...rest}>
        <li className="item-condensed"
        >
          {children ? <span className="material-icons icon-order">arrow_right</span>
            : <span className="icon-order" />}
          {name}
        </li>
      </button>
      {children && showChildren &&
        React.Children.toArray(children).map((_child, _idx) => (
          <div key={_idx} className='tree-element-item'>
            {_child}
          </div>
        ))
      }
    </>
  );
};