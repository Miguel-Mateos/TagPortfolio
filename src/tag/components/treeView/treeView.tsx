import React from 'react';
import { TreeViewElement } from './treeViewElement';

interface ITree {
  /**
   * Alternative className for Tree
   */
  className?: string
  [other: string]: any
}

export const TreeWrap: React.FC<ITree> = ({ children, className, ...rest }) => {
  const renderable = React.Children.toArray(children).every((child: any) => (child.type.name === 'TreeViewElement'));
  return renderable ? (
    <div className={className || ''} {...rest}>{children}</div>
  ) : null;
};

export const Tree = {
  Wrap: TreeWrap,
  Element: TreeViewElement
};