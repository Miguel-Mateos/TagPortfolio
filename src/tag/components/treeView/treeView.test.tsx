import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import { Tree } from './treeView';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const TreeViewTest = (args: any) => {
  return (
    <div className='tag-ds'>
      <Tree.Wrap>
        <Tree.Element data-testid='tree-element-1' name='Item Tree'></Tree.Element>
        <Tree.Element name='Item Tree'></Tree.Element>
        <Tree.Element data-testid='tree-element-2' name='Item Tree'>
          <Tree.Element data-testid='tree-element-3' name='Sub item'>
            <Tree.Element data-testid='tree-element-4' onClick={args.onClick} name='Sub item 2' />
          </Tree.Element>
        </Tree.Element>
      </Tree.Wrap>
    </div>
  );
};

const TreeViewTestNonRenderable = (args: any) => {
  return (
    <div className='tag-ds'>
      <Tree.Wrap>
        <li data-testid='tree-element-1'>item tree</li>
      </Tree.Wrap>
    </div>
  );
};

test('render treeView and match snapshot', () => {
  const component = renderer.create(<TreeViewTest />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('on click function Tree Element', () => {
  const fn = jest.fn();
  const { getByTestId } = render(<TreeViewTest onClick={fn}/>);
  userEvent.click(getByTestId('tree-element-2'));
  userEvent.click(getByTestId('tree-element-3'));
  userEvent.click(getByTestId('tree-element-4'));
  expect(fn).toBeCalledTimes(1);
});

test('non renderable Tree Element', () => {
  const { queryByText } = render(<TreeViewTestNonRenderable />);
  const element = queryByText(/item tree/i);
  expect(element).toBeNull();
});