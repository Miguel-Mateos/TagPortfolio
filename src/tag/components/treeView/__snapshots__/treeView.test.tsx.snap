// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render treeView and match snapshot 1`] = `
<div
  className="tag-ds"
>
  <div
    className=""
  >
    <button
      className="list-icon tree-element "
      data-testid="tree-element-1"
      onClick={[Function]}
    >
      <li
        className="item-condensed"
      >
        <span
          className="icon-order"
        />
        Item Tree
      </li>
    </button>
    <button
      className="list-icon tree-element "
      onClick={[Function]}
    >
      <li
        className="item-condensed"
      >
        <span
          className="icon-order"
        />
        Item Tree
      </li>
    </button>
    <button
      className="list-icon tree-element "
      data-testid="tree-element-2"
      onClick={[Function]}
    >
      <li
        className="item-condensed"
      >
        <span
          className="material-icons icon-order"
        >
          arrow_right
        </span>
        Item Tree
      </li>
    </button>
  </div>
</div>
`;
